command/agent/template/template.go

```diff
@@ -0,0 +1,186 @@
+// Package template is responsible for rendering user supplied templates to
+// disk. The Server type managing the lifecycle of an internal Consul Template
+// Runner
+package template
+
+import (
+	"context"
+	"strings"
+
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/agent/config"
+)
+
+// ServerConfig is a config struct for setting up the basic parts of the
+// Server
+type ServerConfig struct {
+	Logger hclog.Logger
+	// Client        *api.Client
+	VaultConf     *config.Vault
+	ExitAfterAuth bool
+
+	Namespace string
+}
+
+// Server manages the Consul Template Runner which renders templates
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// unblockCh is used to block until a template is rendered
+	unblockCh chan struct{}
+
+	// Templates holds the parsed Consul Templates
+	Templates []*ctconfig.TemplateConfig
+
+	DoneCh        chan struct{}
+	logger        hclog.Logger
+	exitAfterAuth bool
+}
+
+// NewServer returns a new configured server
+func NewServer(conf *ServerConfig) (*Server, chan struct{}) {
+	unblock := make(chan struct{})
+	ts := Server{
+		DoneCh:        make(chan struct{}),
+		logger:        conf.Logger,
+		unblockCh:     unblock,
+		config:        conf,
+		exitAfterAuth: conf.ExitAfterAuth,
+	}
+	return &ts, unblock
+}
+
+// Run kicks off the internal Consul Template runner, and listens for changes to
+// the token from the AuthHandler. If Done() is called on the context, shut down
+// the Runner and return
+func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ctconfig.TemplateConfig) {
+	latestToken := new(string)
+	ts.logger.Info("starting sink server")
+	defer func() {
+		ts.logger.Info("template server stopped")
+		close(ts.DoneCh)
+	}()
+
+	if incoming == nil {
+		panic("incoming channel is nil")
```

Comment:
Shame there's not a way to return an error instead.

Reply:
We could modify this to return an error of course. I was following the precedent of the [SinkServer Run() method][1]. The `AuthHandler` also calls `panic` if an auth method is omitted. At this stage if these things don't exist it's considered an unrecoverable error and we can't continue.


[1]: https://github.com/hashicorp/vault/blob/a1e3c83368e8ae2fe6d54759c24cff53a74ee901/command/agent/sink/sink.go#L96